src/testdrive/src/action.rs

```diff
@@ -392,7 +414,10 @@ pub fn build(cmds: Vec<PosCommand>, state: &State) -> Result<Vec<PosAction>, Err
                         // Skip, has already been handled
                         continue;
                     }
-                    "random-sleep" => Box::new(sleep::build_sleep(builtin).map_err(wrap_err)?),
+                    "random-sleep" => {
+                        Box::new(sleep::build_random_sleep(builtin).map_err(wrap_err)?)
+                    }
+                    "sleep" => Box::new(sleep::build_sleep(builtin).map_err(wrap_err)?),
```

Comment:
I'm going to centralize my response to all the "we shouldn't use sleep" here so that there's only one thread.

> [comment](https://github.com/MaterializeInc/materialize/pull/5723/files/eebc50d017fb71c3b4eaa9c567d969728cd2a8ff..91cca656f4dc7abecd846e927418555101f62864#diff-c10e5b040e85b11a31281a058b791759ae4ce35f601f29be14278f328752e59c)
> This smells like a test flake waiting to happen to me! 

Honestly these tests are _all_ going to be flaky because, per [the AWS docs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/NotificationHowTo.html)

> Typically, event notifications are delivered in seconds but can sometimes take a minute or longer. 

In my testing the current configuration has never failed, but that is just one day of testing. I expect that to make this have 99.9% success we're eventually going to need to add either a `sleep 60` or a `retry for` pragma to specific testdrive queries. Ideally that pragma would also allow for a `fast fail` if we get something that is obviously broken, but that's lower priority.

In addition to the official AWS docs, I can't find any SLA on how long it takes for object notification configurations to go into effect. I worked around it by modifying the configuration early, so that all the retries mean that it's several seconds before we get to the test that actually requires notifications to be configured, but I'm not sure what kind of flake rate it's going to provide us with.

> Is there something we can poll to find out when that source has actually been shutdown? 

The issue is that we need to ensure that the tokio subsystem gets around to polling the reader, which isn't part of the timely/dd infrastructure, it's not even really exposed to the coordinator.

We can make this observable by putting this into an `s3_source_metadata` log or something, which is something I would like to do, but I believe that that would require handing out a reference to the catalog to sources. Which is something we should do, but seems sort of orthogonal.

Reply:
link to an issue for actually providing this metadata: https://github.com/MaterializeInc/materialize/issues/5733

Reply:
Also I think that if this test gets/stays flaky we can move it into disabled until we have all the infrastructure in place to make it a bit more reliable.

Reply:
We actually do have a pragma that can arbitrarily increase the retry frequency for testdrive queries:

https://github.com/MaterializeInc/materialize/blob/747c0bdc6b63a95a94ec7faae693003dc253876e/test/testdrive/testdrive.td#L17

> We can make this observable by putting this into an s3_source_metadata log or something, which is something I would like to do

Yeah, I think having some sort of logging view or metric about that would be great! I saw a thread somewhere about allowing logging from a Tokio task. Definitely going to require some plumbing but should would be nice.

> I believe that that would require handing out a reference to the catalog to sources. Which is something we should do, but seems sort of orthogonal.

Ooh, no, that's very much impossible. The design of the system requires that the catalog be owned by the coordinator at all times. If there is some catalog information that's required it has to be denormalized into a logging view. Of course, those are only eventually consistent... which is why global IDs work the way they do, as in, they are never reused. For the sake of argument, just a simple logging view like

```
CREATE TABLE mz_live_s3_sources (global_id text);
```

would probably be sufficient? Then you could just wait for `SELECT * FROM mz_live_s3_sources WHERE global_id = 'blah'` to return the empty set.

Reply:
Yeah, I guess "hand out a reference to the catalog" was the wrong term, I just meant we need to be able to write to some metadata tables from sources, which is something we don't currently have the infra to do.

That specific table wouldn't be sufficient, because we need to wait for all the the S3 source ingestors to be shut down, but something like it would be.